### PR TITLE
Make bandit module public

### PR DESCRIPTION
**Pull request by MaxHalford**
Ok, here's a rather big PR to make the `bandit` module public.

Until now, we had some bandit method in the `_bandit` method. We used those in `model_selection` to implement model selection based on bandit logic.

I refactored a bunch of stuff to make the bandits generic. They should be usable in all sorts of situation, with model selection being a special case. In particular, the `reward_obj` within each bandit can be a `proba.base.Distribution`, a `stats.base.Statistic`, or a `metrics.base.Metric`. To make this possible, I made it so that each one of these implements a `>` operator to compare two instances. That was pretty straightforward for metrics and statistics because both have a `get` method. For distribution I added a `mode` property to each distribution.

I also added a (tested) gym environment to reproduce the Candy Cane Competition from Kaggle. Right now it only works for one player, but that can be changed in a future PR.

I added logic to unit test all the (policies x environments) to make sure each policy performs better than a random policy.

The next big steps are:

- Implement a multiplayer version of the candy cane contest, and show how to use bandits in an adversarial situation.
- Implement a contextual bandit, such as LinUCB

I'll add these to the roadmap :)

@reviewers I don't expect you to review all the code. It would be just be cool if you can skim over the bandit API and raise any flags if you see any. Cheers!

### CODE REVIEW

**Comment by smastelini** [comment]
Nice work @MaxHalford ! I can give it a look tomorrow. If you are happy with it and in a hurry you can merge and any problems we might find can be fixed later :)

**Comment by MaxHalford** [comment]
There's no hurry at all :)

File: river/bandit/thompson.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    @property
+    def dist(self):
+        return self.reward_obj
+
+    def _pull(self, arms):
+        return max(arms, key=lambda arm: self._rewards[arm].sample())
+
+    @classmethod
+    def _unit_test_params(cls):
+        yield {"dist": proba.Beta()}
```
Comment (AdilZouitine) [inline river/bandit/thompson.py:86]: I found it confusing to have a private class method only useful for unit testing. Is there an alternative?

File: river/bandit/ucb.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            )
+            if (reward := self._rewards.get(arm)) is not None
+            else math.inf
+            for arm in arms
+        }
+        return max(arms, key=lambda arm: upper_bounds[arm])
+
+    @classmethod
+    def _unit_test_params(cls):
+        yield {"delta": 1}
```
Comment (AdilZouitine) [inline river/bandit/ucb.py:80]: Same comment as other bandits

File: river/metrics/base.py [hunk 3]
```diff
@@ -52,6 +52,9 @@ def is_better_than(self, other) -> bool:
         op = operator.gt if self.bigger_is_better else operator.lt
         return op(self.get(), other.get())
 
+    def __gt__(self, other):
+        return self.is_better_than(other)
+
```
Comment (AdilZouitine) [inline river/metrics/base.py:57]: Love it a lot!

File: river/model_selection/bandit.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        return self
+
+    @classmethod
+    def _unit_test_params(cls):
+        for params in super()._unit_test_params():
+            yield {**params, "policy": bandit.EpsilonGreedy(0.2)}
+
+    def _unit_test_skips(self):
+        return (
+            {"check_model_selection_order_does_not_matter"} if hasattr(self.policy, "seed") else {}
```
Comment (AdilZouitine) [inline river/model_selection/bandit.py:112]: Same comment as the bandit unit testing class method

File: setup.py [hunk 5]
```diff
@@ -57,6 +57,7 @@
             "black>=22.1.0",
             "flake8>=4.0.1",
             "graphviz>=0.10.1",
+            "gym>=0.26.1",
```
Comment (AdilZouitine) [inline setup.py:60]: Gym has been quite unstable for the last couple of months. We should freeze the version for the following month to wait for the gym team to stabilize the API.

**Review by AdilZouitine — COMMENTED** [review]
It's a great pull request, and I think we can quickly expand it for multiple players. 
I made some minor remarks!
I can't wait to see this pull request merge 😄

File: setup.py (hunk 5 above)
Comment (MaxHalford) [reply setup.py:60]: Sure! What version do you suggest?

File: river/bandit/base.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    ----------
+    reward_obj
+        The reward object used to measure the performance of each arm. This can be a metric, a
+        statistic, or a distribution.
+    burn_in
+        The number of steps to use for the burn-in phase. Each arm is given the chance to be pulled
+        during the burn-in phase. This is useful to mitigate selection bias.
+    seed
+        Random number generator seed for reproducibility.
+
```
[the remark this answers is not in the document]
Comment (MaxHalford) [reply river/bandit/base.py]: Good catch

File: river/bandit/base.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+RewardObj = Union[stats.base.Statistic, metrics.base.Metric, proba.base.Distribution]
+
+
+class Policy(base.Base, abc.ABC):
+    """Bandit policy base class.
+
+    Parameters
+    ----------
+    reward_obj
+        The reward object used to measure the performance of each arm. This can be a metric, a
```
[the remark this answers is not in the document]
Comment (MaxHalford) [reply river/bandit/base.py:19]: I think so too! A nice trick is that you pass a rolling object if you want your bandit to handle drift. I'll document that at some point :)

File: river/bandit/epsilon_greedy.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    def _pull(self, arms):
+        return (
+            self._rng.choice(arms)  # explore
+            if self.best_arm is None or self._rng.random() < self.current_epsilon
+            else self.best_arm  # exploit
+        )
+
+    @classmethod
+    def _unit_test_params(cls):
+        yield {"epsilon": 0.2}
```
[the remark this answers is not in the document]
Comment (MaxHalford) [reply river/bandit/epsilon_greedy.py:87]: Why is that?

We need some mechanism to specify unit test params and unit tests to skip. Would you rather define them all in a single file? I did this it this way following scikit-learn, which does this for skipping unit tests.

**Review by gbolmier — COMMENTED** [review]
I got confused by the `arm` naming convention because it actually is the arm "id", "index", "label" (or whatever we want to call it) and not the arm itself. Otherwise LGTM, great work!

File: setup.py (hunk 5 above)
Comment (gbolmier) [reply setup.py:60]: Let's cap and not freeze then @AdilZouitine 👍

**Comment by MaxHalford** [comment]
Yes good point @gbolmier. I added a type hint to make this a bit clearer

**Comment by gbolmier** [comment]
Yeah, the type hint actually saved my sanity lol

> Would you rather call it `arm_index`?

That would be clearer IMO. I'm fine with `arm_index` but why not `arm_id` or `arm_ix` if people find the former too verbose.

**Comment by MaxHalford** [comment]
I'll think about it. The thing about `index` is that it implies (to me) that the value is an integer. But the arm can also be a `str`. In fact, the arm can be anything hashable.

**Comment by gbolmier** [comment]
Yep, that's why I mentioned `arm_id`.

File: river/bandit/thompson.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+    """
+
+    def __init__(self, dist: proba.base.Distribution, burn_in=0, seed: int = None):
+        super().__init__(dist, burn_in)
+        self.seed = seed
+        self._rng = random.Random(seed)
+        self._rewards.default_factory = self._clone_dist_with_seed
+
+    def _clone_dist_with_seed(self):
```
Comment (smastelini) [inline river/bandit/thompson.py:74]: I did that in the Periodic drift detector :D

Comment (smastelini) [reply river/bandit/thompson.py:74]: Soft gambiarra ;)

**Comment by smastelini** [comment]
I don't know much about bandits and reinforcement learning, so I'll leave the technical bit for the experts :)

But design-wise, LGTM! I really like the `BanditRegressor` class and I think a lot of people will like that too.

File: setup.py (hunk 5 above)
Comment (AdilZouitine) [reply setup.py:60]: Yes I think is a good option!

**Comment by AdilZouitine** [comment]
> Ok, I took into account your suggestion, @gbolmier. You're right I think it looks better :)
> 
> @AdilZouitine feel free to suggest if you come up with a better design for test default parameters, we can address it in a future PR :)

Yes, I'll think about it 😄

**Pull request merged**